File: [0338][Medium][Counting_Bits]/Counting_Bits.py
# leetcode time     cost : 96 ms
# leetcode memory   cost : 14.6 MB 
# Time  Complexity: O(n*k)
# Space Complexity: O(n)
# slolution 1, num_i &= (num_i -1)
class Solution:
    def countBits(self, num: int) -> List[int]:
        result = []
        for num_i in range(num+1):
            count = 0
            while num_i:
                count+=1
                num_i &= (num_i -1)
            result.append(count)
        return result

File: [0338][Medium][Counting_Bits]/Counting_Bits.java
// leetcode time     cost : 2 ms
// leetcode memory   cost : 43 MB 
// Time  Complexity: O(n*k)
// Space Complexity: O(n)
// slolution 1, num_i &= (num_i -1)
public class Solution {
    public int[] countBits(int num) {
        int[] ans = new int[num + 1];
        for (int i = 0; i <= num; ++i)
            ans[i] = popcount(i);
        return ans;
    }
    private int popcount(int x) {
        int count;
        for (count = 0; x != 0; ++count)
          x &= x - 1; //zeroing out the least significant nonzero bit
        return count;
    }
}